refactor: load firebase_admin lazily

jonesnc opened the pull request:
## Changes
Lazily import `firebase_admin` and its modules because we've found it's one of the slower modules to import in our Django projects. For type annotations, the `firebase_admin` is only imported after checking `typing.TYPE_CHECKING`.

Please let me know if you have any suggestions or feedback on this, I'd be happy to implement them.

Thanks for this package, I think it's great!

## Tests
```bash
Test session starts (platform: linux, Python 3.10.12, pytest 8.2.2, pytest-sugar 1.0.0)
django: version: 4.2.19, settings: tests.settings.default (from ini)
rootdir: /home/nathanjones/Projects/fcm-django
configfile: pyproject.toml
plugins: sugar-1.0.0, anyio-3.6.2, mock-3.14.1, django-4.11.1, asyncio-0.14.0, Faker-15.3.4, typeguard-2.13.3
collected 16 items

 tests/test_admin.py ✓✓                                                                                                                                                                                                                      12% █▍
 tests/test_api_rest_framework.py ✓✓✓                                                                                                                                                                                                        31% ███▎
 tests/test_api_tastypie.py ✓                                                                                                                                                                                                                38% ███▊
 tests/test_models.py ✓✓✓✓✓✓✓✓✓✓                                                                                                                                                                                                            100% ██████████

Results (3.49s):
      16 passed

```

jonesnc committed:
refactor: load firebase_admin lazily

jonesnc committed:
style: fix formatting issues

ahmedbilal commented:
@jonesnc This latest release 2.3.1 bump up the ceiling for `firebase_admin` constraint allowing [firebase_admin 7](https://github.com/firebase/firebase-admin-python/releases/tag/v7.0.0) as dependency which removed the dependency on `google-api-python-client` thus significantly reducing the overall footprint (memory + size) of this package. Can you run your profiling again and see whether you need these changes? Thanks.

jonesnc commented:
It looks like my project will take some work to update to `firebase_admin` version 7, so I can't provide any `importtime` benchmark results at this time. I'll work on this and get back with you.

jonesnc commented:
@ahmedbilal I finally did some testing by upgrading to version 2.3.1, and found that while removing the `googleapiclient` import saved ~*35*ms import time (which is appreciated), the `firebase_admin` package import is still taking up ~*134*ms import time. I'd still consider that a win in my book.

I'll go ahead and fix the branch conflicts.

jonesnc committed:
feat: add lazy Firebase app initialization

Add INITIALIZE_APP_CALLABLE setting that defers Firebase
initialization until first use. The callable is invoked
lazily and the result is cached for subsequent calls.

jonesnc committed:
docs: add INITIALIZE_APP_CALLABLE to README

jonesnc committed:
docs: expand INITIALIZE_APP_CALLABLE examples

xtrinch committed:
Refactor lazy Firebase access

xtrinch committed:
Clean up lazy Firebase typing

xtrinch committed:
Rename Firebase app initializer setting

xtrinch committed:
Fix lint formatting

xtrinch committed:
test: fix lazy firebase rebase follow-ups

ahmedbilal commented:
@jonesnc @xtrinch Not sure whether this would help this cause or not, but see https://docs.python.org/3.15/whatsnew/3.15.html#pep-810-explicit-lazy-imports

xtrinch commented:
@ahmedbilal This will only be useful once Python ≥3.15 becomes the baseline (e.g. in LTS environments). Right now, it doesn’t help us.

jonesnc commented:
I'm happy to just wait until Oct 1, 2026, when Python 3.15 is released. Would that be a good time to revisit this, or do you want to wait for a later date?

xtrinch commented:
@jonesnc Yes, that works, but we'd need to make sure it still works for earlier python versions, even if make lazy imports work for python 3.15.

jonesnc commented:
@xtrinch From [PEP 810](https://docs.python.org/3.15/whatsnew/3.15.html#pep-810-explicit-lazy-imports):

> For code that cannot use the lazy keyword directly (for example, when supporting Python versions older than 3.15 while still using lazy imports on 3.15+), a module can define [__lazy_modules__](https://docs.python.org/3.15/reference/datamodel.html#module.__lazy_modules__) as a container of fully qualified module name strings. Regular import statements for those modules are then treated as lazy, with the same semantics as the lazy keyword:

```python
__lazy_modules__ = ["json", "pathlib"]

import json     # lazy
import os       # still eager
```

So we could use `__lazy_modules__` instead of the `lazy import` syntax, which would be compatible with previous python versions.
